### DISCUSSION TITLE
Improve `should_save` and potentially add `should_start`?

### DISCUSSION THREAD

**Discussion by zhming0** [olimorris/persisted.nvim#153] [Ideas]
I wonder if this is helpful to make a default `should_save` like this? Transient buffers don't feel valuable to keep. 

Also, I noticed that `should_save` does not evaluate each buffer, it only evaluate the single buffer when we exist Neovim (VimLeavePre), is it better to change it to evaluate each buffer filtering out unwanted buffers? 

```lua
      -- Proposed default should_save
      ---@return boolean
      should_save = function()
        if vim.bo.buftype ~= "" then
          -- we don't want things like quickfix etc to be saved
          return false
        end
        return true
      end,
```

Lastly, I wonder if a configurable `should_start` function will be useful? We can use it to implement things like "only start in a directory with .git".

**Comment by olimorris** [comment]
I'm not going to add any more methods to the plugin for now, as per #145. I'm come to realise that every Neovim user has specific requirements when it comes to session management and it's impossible to cater for all. So I've done by best to make the plugin extensible.

**Reply by zhming0** [reply]
Thank you for replying. I understand. Just wanting to clarify the intention of `should_save` though: reading the readme, I feel you wanted it to "check whether each buffer should be saved", but the current behavior isn't like that.

Am I right? Or is the current behavior intended?

Right now my workaround is: 

```lua
    vim.api.nvim_create_autocmd("User", {
      pattern = "PersistedSavePre",
      callback = function()
        local buffers = vim.api.nvim_list_bufs()

        for _, buf in ipairs(buffers) do
          if vim.api.nvim_buf_is_loaded(buf) then
            local buftype = vim.api.nvim_buf_get_option(buf, "buftype")
            local modifiable = vim.api.nvim_buf_get_option(buf, "modifiable")

            if buftype ~= "" or not modifiable then
              vim.api.nvim_buf_delete(buf, { force = true })
            end
          end
        end
      end,
    })
```

**Reply by olimorris** [reply]
If I recall, I added it because there were times when I'd load Neovim, hit my Alpha dashboard, then be hit with an error from a plugin that I'd just updated. I'd quit Neovim, sort the error, load Neovim again and be presented with a blank buffer as my session had been saved when I was in the dashboard.

So `should_save` was a simple way of checking whether you the session should save given the present state of Neovim.

Your solution has made me realise that this should really all be handled via `PersistedSavePre`, doing away with `should_save`.

Your solution seems absolutely correct for your use case. You could technically put it in `should_save` and return a boolean.
